### ISSUE TITLE
Update host logic to support optional VM assembly

### ISSUE THREAD

**Issue by marek-safar** [dotnet/runtime#47481]
For desktop scenarios on architectures where CoreCLR is not supported we need to use MonoVM as supported runtime. Today the host hardcodes libcoreclr name as the only one to look for which makes this problematic because MonoVM capabilities and internals are different.

This should unblock issues like #34202.

/cc @steveisok

**Comment by marek-safar** [comment]
ABI would be the same as for coreclr and this is for shared framework scenarios where CoreCLR is not supported like s390x

**Comment by vitek-karas** [comment]
I see - do we need to case where on one machine I have shared frameworks with both CoreCLR and MonoVM? (I hope not, otherwise framework resolution will become more complex than it already is).

**Comment by marek-safar** [comment]
> do we need to case where on one machine I have shared frameworks with both CoreCLR and MonoVM?

No, the simple case should be enough at least for now (although it would help with development/testing)

**Comment by vitek-karas** [comment]
In that case we can build a different version of `hostpolicy` for these "mono" platforms.
/cc @VSadov - as this might affect the work to move hosting components from the installer subtree to some other place

**Comment by akoeplinger** [comment]
I think we need some kind of runtime switch for choosing between mono/coreclr on the standard platforms too if we want to really unblock https://github.com/dotnet/runtime/issues/34202 since right now we need to rename libmono to libcoreclr in the testhost for libraries testing.

**Comment by steveisok** [comment]
It would be nice if we could have something like `--vm mono` as a runtime option and be able to switch between coreclr / mono in the same install.  Assuming we have different names between libmono and libcoreclr, I *think* we would just have to put each runtime's System.Private.CoreLib in a separate directory.  Not sure how easy that would be.

**Comment by vitek-karas** [comment]
@steveisok this has been discussed in the past. It's definitely possible, but it's a rather large change as it would require changes to runtime packs, SDK, installers, ... There's also the question of breaking apps due to files lying in a different locations on disk (the typically `typeof(object).Assembly.Location` would now point to a subdirectory) and so on.

If we were to support installing both CoreCLR and MonoVM shared frameworks on the same machine, then the best solution would be to include some differentiator elsewhere. The problem is that currently the system doesn't have a place to do this - adding another level of directories will probably break too many things if we do this. And it still impacts pretty all parts of the toolchain.

If this would be only for development/test purposed, then I think the easiest way would be to simply create a new installation in a different location and rely on `DOTNET_ROOT`. This is already done in many places to handle different versions and such, and so it has broad support across our tooling. The downside is size on disk - it would have to be a full install of at least the runtime (so all libraries + runtime).

**Comment by akoeplinger** [comment]
We don't need to change the location of coreclr, we could add a `mono/` subfolder and put libmono + System.Private.Corelib there (though not sure how well supported having SPC in a different folder from the rest of the BCL is).

> If this would be only for development/test purposed, then I think the easiest way would be to simply create a new installation in a different location and rely on DOTNET_ROOT

We'd still need some switch to tell the host to look for libmono instead. Or we could probe for `libcoreclr` and fallback to `libmono` if it doesn't exist? For architectures like s390x the libcoreclr check could be ifdefed out.

**Comment by vitek-karas** [comment]
I think it's perfectly OK to accept more than one name for the runtime itself - so the fallback logic could definitely work. There will be another small gotcha - currently the host requires coreclr.dll to be listed in `.deps.json` - so to produce a valid MonoVM based framework it would now have to use the `mono.dll` in the `.deps.json` instead. It's definitely the cleaner way.

Just so that this is actually actionable - some clarifications:
* What is the name of the mono runtime library we want the host to recognize? `mono` (plus all the required platform prefixes and suffixes).
* Do we need any changes to the ABI through which `hostpolicy` communicates with the runtime library? (`coreclr_initialize` and so on)

**Comment by jkotas** [comment]
What would be in the .deps.json that the build produces?

I like the idea of building two different `hostpolicy` .dlls and having either coreclr-based shared runtime or mono-based shared runtime (but not both) much better.

**Comment by akoeplinger** [comment]
> so to produce a valid MonoVM based framework it would now have to use the mono.dll in the .deps.json instead. It's definitely the cleaner way.

As far as I know the .deps.json is generated based on the contents of the runtime pack so it should automatically get the libmono instead.

> What is the name of the mono runtime library we want the host to recognize? mono (plus all the required platform prefixes and suffixes).

Yes.

> Do we need any changes to the ABI through which hostpolicy communicates with the runtime library? (coreclr_initialize and so on)

I don't think so, we implement these right now: https://github.com/dotnet/runtime/blob/e9695b8e7840c37390ca58b821cfbeab178d1cc6/src/mono/mono/mini/main-core.c#L29-L41

**Comment by vitek-karas** [comment]
Re `.deps.json`:
The one in self-contained apps is generated from the runtime pack. But large part of this discussion is to support framework-dependent apps on mono VM. So the question is how are we going to produce that:
* Microsoft.NETCore.App (currently built by the installer subset in `dotnet/runtime` would need to get a "mono" variant
* It's probably easy enough for platforms which don't support CoreCLR - there we simply include MonoVM instead and ship it
* How will this work for platforms which do support CoreCLR? This is a similar/related question to how are we differentiating the runtime packs between CoreCLR/MonoVM for RIDs which support both.

**Comment by akoeplinger** [comment]
`.deps.json` for framework-dependent apps doesn't contain a reference to libcoreclr so that should not need any changes.

We already build variants of the runtime pack for platforms also supported by CoreCLR, there's an additional `.Mono` in the package name: https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet6&package=Microsoft.NETCore.App.Runtime.Mono.linux-x64&protocolType=NuGet&version=6.0.0-preview.2.21078.3

Consuming this is also already supported by the SDK since https://github.com/dotnet/sdk/pull/11824

**Comment by vitek-karas** [comment]
The problem is with the framework itself - it also contains `.deps.json` (specifically `Microsoft.NETCore.App.deps.json`) and that contains `libcoreclr.so` - you can also see this in the runtime pack which has this file it as well (and I think it will need to be modified for self-contained apps to use `libmono` instead.

The runtime packs solve the self-contained apps, for framework-dependent apps we would need to also produce at least `.tar.gzip`/`.zip` files (for installation to a private location and then using the `DOTNET_ROOT` as mentioned above).

**Comment by akoeplinger** [comment]
Not sure I get what the problem with the .deps.json is, a mono-based shared framework like the one I linked lists libcoreclr.so but only because we renamed the file. If we rename it back to libmono.so then the .deps.json will be correct too.

**Comment by vitek-karas** [comment]
Good - in that case it should work by just teaching `hostpolicy` about the different name.

**Comment by vitek-karas** [comment]
If https://github.com/dotnet/runtime/issues/34202 should be done in 6, then this needs to be done as well. Otherwise we should move both to future release.

**Comment by akoeplinger** [comment]
I do think this should be done for 6, we just need to agree on the approach. I think given the timeframe I prefer probing for libcoreclr and falling back to libmono if it doesn't exist, that should be a very small change to the host.

**Comment by jkotas** [comment]
@akoeplinger Why do we need to do it for 6 without #34202 ? It would be future proofing that rarely works in practice.

**Comment by akoeplinger** [comment]
@jkotas my understanding is we should get https://github.com/dotnet/runtime/issues/34202 done in 6.0 too, but it depends on this issue.

**Comment by jkotas** [comment]
Ah ok, I got confused between #34202 and #49661 (should one be resolve as duplicate of the other?).  #49661 was moved to .NET 7.

**Comment by akoeplinger** [comment]
Yes, I closed https://github.com/dotnet/runtime/issues/49661 as a duplicate.

**Comment by agocke** [comment]
Is this still an issue? I thought Mono was currently successfully using the host in non-mobile scenarios.

**Comment by steveisok** [comment]
This is still an issue.  We want to have the host recognize a vm lib other than libcoreclr.  I don't think we'll get to this anytime soon.

**Comment by agocke** [comment]
I'll add it to 8.0, since this will likely not make it for 7.0.

**Comment by akoeplinger** [comment]
Closing this since we'll probably not going to implement this.
